pkg/controller/data_upload_controller.go

```diff
@@ -0,0 +1,498 @@
+	csiSnapshotClient snapshotter.SnapshotV1Interface, repoEnsurer *repository.Ensurer, clock clocks.WithTickerAndDelayedExecution,
+	cred *credentials.CredentialGetter, nodeName string, fs filesystem.Interface, log logrus.FieldLogger) *DataUploadReconciler {
+	return &DataUploadReconciler{
+		client:              client,
+		kubeClient:          kubeClient,
+		csiSnapshotClient:   csiSnapshotClient,
+		clock:               clock,
+		credentialGetter:    cred,
+		nodeName:            nodeName,
+		fileSystem:          fs,
+		logger:              log,
+		repoEnsurer:         repoEnsurer,
+		snapshotExposerList: map[velerov2alpha1api.SnapshotType]exposer.SnapshotExposer{velerov2alpha1api.SnapshotTypeCSI: exposer.NewCSISnapshotExposer(kubeClient, csiSnapshotClient, log)},
+		dataPathMgr:         datapath.NewManager(1),
+	}
+}
+
+// +kubebuilder:rbac:groups=velero.io,resources=dataupload,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=velero.io,resources=dataupload/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get
+// +kubebuilder:rbac:groups="",resources=persistentvolumes,verbs=get
+// +kubebuilder:rbac:groups="",resources=persistentvolumerclaims,verbs=get
+func (r *DataUploadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.logger.WithFields(logrus.Fields{
+		"controller": "dataupload",
+		"dataupload": req.NamespacedName,
+	})
+	log.Infof("Reconcile %s", req.Name)
+	var du velerov2alpha1api.DataUpload
+	if err := r.client.Get(ctx, req.NamespacedName, &du); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debug("Unable to find DataUpload")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrap(err, "getting DataUpload")
+	}
+
+	if du.Spec.DataMover != "" && du.Spec.DataMover != dataMoverType {
+		log.WithField("Data mover", du.Spec.DataMover).Info("it is not one built-in data mover which is not supported by Velero")
+		return ctrl.Result{}, nil
+	}
+
+	ep, ok := r.snapshotExposerList[du.Spec.SnapshotType]
+	if !ok {
+		return ctrl.Result{}, fmt.Errorf("%v type of snapshot exposer is not exist", du.Spec.SnapshotType)
+	}
+
+	if du.Status.Phase == "" || du.Status.Phase == velerov2alpha1api.DataUploadPhaseNew {
+		log.Info("Data upload starting")
+
+		accepted, err := r.acceptDataUpload(ctx, &du)
+		if err != nil {
+			return r.errorOut(ctx, &du, err, "error to accept the data upload", log)
+		}
+
+		if !accepted {
+			log.Debug("Data upload is not accepted")
+			return ctrl.Result{}, nil
+		}
+
+		log.Info("Data upload is accepted")
+
+		exposeParam := r.setupExposeParam(&du)
+		if err := ep.Expose(ctx, getOwnerObject(&du), du.Spec.CSISnapshot.VolumeSnapshot, du.Spec.OperationTimeout.Duration, exposeParam); err != nil {
+			return r.errorOut(ctx, &du, err, "error to expose snapshot", log)
+		}
+		log.Info("Snapshot is exposed")
+
+		return ctrl.Result{}, nil
```

Comment:
To me, `ep.Expose()` is a sync operation. After it returns, everything is already done, so it seems no need to return here. Just moving forward should be OK.

Reply:
`ep.Expose()` will trigger to create one pod whose volume is restored by a given volume snapshot, but the pod maybe is not in the same node of the current controller, so we need to return here.
And then only the controller who is in the same node could do the rest work.

Reply:
Understood, then please also add some comments to explain the reason.

Reply:
Added